clients/src/main/java/org/apache/kafka/clients/producer/internals/IncompleteBatches.java

```diff
@@ -51,6 +51,14 @@ public void remove(ProducerBatch batch) {
         }
     }
 
+    public Iterable<ProduceRequestResult> requestResults() {
+        synchronized (incomplete) {
+            ArrayList<ProduceRequestResult> results = new ArrayList<>(this.incomplete.size());
+            this.incomplete.forEach(incomplete -> results.add(incomplete.produceFuture));
+            return results;
```

Comment:
Out of curiosity, why did you not use `map` here? Is it to avoid the allocation of the `stream`? It doesn't seem that this is a hotspot though.

Reply:
I think I had a bit of a blind spot because I wanted to preallocate the ArrayList to be the correct size. I think streams is able to do that though. I'm happy to convert it back if it's nicer.

Reply:
Let's do it. :)

Reply:
Done.